tests/e2e/attention/generate_e2e_fa2_tests.py

```diff
@@ -0,0 +1,516 @@
+    )
+    result_tensor_type = f"tensor<{result_shape[0]}x{result_shape[1]}x{result_shape[2]}x{value_type.value}>"
+    F32 = "f32"
+    F16 = "f16"
+    op_name = "iree_linalg_ext.attention"
+
+    # Compilation info is optional; prints empty string by default.
+    func_definition = ""
+
+    signature = f"({query_tensor_type}, {key_tensor_type}, {value_tensor_type}, {result_tensor_type}) -> {result_tensor_type}"
+    import_declaration = f"func.func private @module.{func_name}(%query: !hal.buffer_view, %key: !hal.buffer_view, %value: !hal.buffer_view, %scale: f32) -> !hal.buffer_view"
+    func_definition = func_definition + (
+        f"func.func @{func_name}(%query: {query_tensor_type}, %key: {key_tensor_type}, %value: {value_tensor_type}, %scale: {F32}) -> {result_tensor_type} {{\n"
+        f"  %result0 = tensor.empty(): {result_tensor_type}\n"
+        f"  %scale_f16 = arith.truncf %scale : f32 to f16 \n"
+        f"  %result1 = {op_name} {{\n"
+        f"      indexing_maps = [affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d2)>,\n"
+        f"                       affine_map<(d0, d1, d2, d3, d4) -> (d0, d3, d2)>,\n"
+        f"                       affine_map<(d0, d1, d2, d3, d4) -> (d0, d3, d4)>,\n"
+        f"                       affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d4)>]\n}}"
+        f"      ins(%query, %key, %value, %scale_f16: {query_tensor_type}, {key_tensor_type}, {value_tensor_type}, {F16})\n"
+        f"      outs(%result0: {result_tensor_type}) -> {result_tensor_type}\n"
+        f" return %result1: {result_tensor_type}\n"
+        f"}}\n"
+    )
+    return MLIRFunction(
+        name=func_name,
+        signature=signature,
+        import_declaration=import_declaration,
+        definition=func_definition,
+    )
+
+
+# Represents a call to a generated test function.
+@dataclasses.dataclass
+class TestCall:
+    function: MLIRFunction
+    op: str
+
+
+# Enumerates ways to initialize tensor buffer contents.
+@enum.unique
+class TensorGenerator(enum.Enum):
+    ZERO = "zero"  # Fill with zeros
+    RANDOM = "random"  # Fill with (deterministic) pseudorandom values.
+
+
+# Intentionally fixed seed! We want full reproducibility here, both across runs
+# and across machines.
+# Intentionally not shared with local_pseudorandom_state to limit the ways
+# in which shuffling testcases changes which random values are generated.
+pseudorandom_generator_seed = 1
+
+
+def contents_generator_tag(generator: TensorGenerator):
+    if generator == TensorGenerator.ZERO:
+        return ""
+    elif generator == TensorGenerator.RANDOM:
+        global pseudorandom_generator_seed
+        pseudorandom_generator_seed = pseudorandom_generator_seed + 1
+        return f"!tag:iree:fully_specified_pseudorandom {pseudorandom_generator_seed}"
+    else:
+        raise ValueError(generator)
+
+
+# Generate a 3d tensor function argument of the given size as `%name`.
+def generate_random_3d_tensor(
```

Comment:
I thought we were generating 4d tensors.

Reply:
I was thinking the same. I synced with @raikonenfnu for that. Based on the mapping below, we have 3D tensors:
Q_map = (b, m, k1, k2, n) -> (b, m, k1)
K_map = (b, m, k1, k2, n) -> (b, k2,k1)
V_map (b, m, k1, k2,n) -> (b, k2, n)
O_map (b, m, k1, k2, n) -> (b, m, n)